### DISCUSSION TITLE
Add Watermark to Spatie Image

### DISCUSSION THREAD

**Discussion by PrashantDhungana** [filamentphp/filament#7446] [Help] [answered]
I am using Spatie Media Library plugin for filament. I have to add an image watermark on top of the uploaded image. I checked if the image comes in `mutateFormDataBeforeCreate(array data)` function, but the data doesn't contain the image. So how do i manipulate the specific image before saving it?

**Comment by danharrin** [comment]
Maybe an observer for the Media model?

**Reply by PrashantDhungana** [reply, marked as answer]
Thanks a lot @danharrin . Was focused on filament that I missed the Laravel and Spatie way. While solving this problem, Since the Media model was used for several of my other Models. But I only wanted to add the watermark in one specific model. 

I solved it using spatie's `registerMediaConversions()` function to add watermark for the image. 

In case, if any future devs go through the same problem, here is how I solved it :

`public function registerMediaConversions(Media $media = null) : void
    {
        $imageSize = getimagesize(public_path("/storage/$media->id/$media->file_name"));

        $watermark = $this->createWatermarkConversion($imageSize[0], $imageSize[1]);
        // Watermark conversion
        $this->addMediaConversion('watermark')
            ->watermark($watermark);

    }

    protected function createWatermarkConversion($width, $height)
    {
        $img_canvas = Image::canvas($width, $height);
        
        $stitched = uniqid('watermark_');
        $img_canvas->insert(public_path('/images/watermark.png'));
        $img_canvas->insert(public_path('/images/watermark.png'), 'top-left',3875);
        $img_canvas->save(public_path("/images/$stitched.png"), 100);

        return public_path("/images/$stitched.png");
    }`

**Reply by Webmitr** [reply]
can you guide on this a bit more...?

**Reply by kvachakhia** [reply]
> Thanks a lot @danharrin . Was focused on filament that I missed the Laravel and Spatie way. While solving this problem, Since the Media model was used for several of my other Models. But I only wanted to add the watermark in one specific model.
> 
> I solved it using spatie's `registerMediaConversions()` function to add watermark for the image.
> 
> In case, if any future devs go through the same problem, here is how I solved it :
> 
> `public function registerMediaConversions(Media $media = null) : void { $imageSize = getimagesize(public_path("/storage/$media->id/$media->file_name"));
> 
> ```
>     $watermark = $this->createWatermarkConversion($imageSize[0], $imageSize[1]);
>     // Watermark conversion
>     $this->addMediaConversion('watermark')
>         ->watermark($watermark);
> 
> }
> 
> protected function createWatermarkConversion($width, $height)
> {
>     $img_canvas = Image::canvas($width, $height);
>     
>     $stitched = uniqid('watermark_');
>     $img_canvas->insert(public_path('/images/watermark.png'));
>     $img_canvas->insert(public_path('/images/watermark.png'), 'top-left',3875);
>     $img_canvas->save(public_path("/images/$stitched.png"), 100);
> 
>     return public_path("/images/$stitched.png");
> }`
> ```

Need better explain what need to do
